### PR TITLE
Derive enum

### PR DESCRIPTION
**Pull request by b05902132**
This PR is about #834. Actually it's just #1045 rebased on the current main branch.

However, due to PyO3 internal API changes, I have to rewrite how `pyenum` implements PyO3 traits, and I am not sure whether I got it right. I am sure I omitted docstrings and descriptors, but maybe there are some other traits that I need to implement, or maybe I implemented something wrong.

The origin implement tries to create a subclass for each variant. [This isn't compatible with current PyO3 API](https://github.com/PyO3/pyo3/issues/417#issuecomment-971767941), because we don't have a way to convert the subclass back to the enum.  To make PyO3 support this is beyond my capability, so this PR will use a simpler approach. But to support #417 and #1714 we will eventually have to do that.

This PR will only support enums with unit variants, by implement the enum as a `PyClass` with an `__eq__` method, and add a `classattr` for each variant.

Unsolved issues:
  - ~~How should we treat custom discriminant values? (Add an `__int__` implementation?)~~ Will not touch it at this PR.
  - How should we implement `__repr__`?
  - How does docstring on each variant work?

### CODE REVIEW

File: src/types/enum.rs
Comment (davidhewitt) [inline src/types/enum.rs:1]: I think this file is dead code which can be thrown away

File: pyo3-macros-backend/src/pyenum.rs
Comment (davidhewitt) [inline pyo3-macros-backend/src/pyenum.rs:76]: This is clever, but I think the downside of this is that without `multiple-pymethods` feature it won't be possible for users to add their own `pymethods` block.

You should instead use [`gen_py_const`](https://github.com/PyO3/pyo3/blob/2f6ea2f96cf8d9f213ea6c8b047e95d5e4aecaca/pyo3-macros-backend/src/pyimpl.rs#L121) to create "method definitions" which then need to be fed into the collection machinery via a "methods trait" - probably `PyClassDescriptors` can be "abused" here (that's the one which collects all the `#[pyo3(get, set)]` annotations from regular `#[pyclass]`).

Let me know if you want me to give some more pointers - this code is rather messy and not well documented as it's in a permanent state of evolution / hackery :)

File: pyo3-macros-backend/src/pyenum.rs
Comment (davidhewitt) [inline pyo3-macros-backend/src/pyenum.rs:48]: Does there need to be separate classes for the variants? I understood that was not the plan any more, maybe this is code to be killed off?

File: pyo3-macros-backend/src/pyimpl.rs
Comment (davidhewitt) [inline pyo3-macros-backend/src/pyimpl.rs:37]: 👍 this builder class is a nice refactoring!

File: tests/test_enum.rs
Comment (davidhewitt) [inline tests/test_enum.rs:8]: Crazy idea, does this need to be a separate `#[pyenum]` macro?

This still creates a Python class, so maybe this can just be a new feature for the `#[pyclass]` macro?

**Review by davidhewitt — COMMENTED** [review]
Thank you, I'm really excited to see progress being made for this again!

I've posted a few comments to get started. I think that may lead to a fair few changes so I've held off on more detailed review for now.

File: tests/test_enum.rs
Comment (b05902132) [reply tests/test_enum.rs:8]: I think that's a great idea. I'm not sure how do that interact with [https://pyo3.rs/v0.15.0/class.html#customizing-the-class] though. Maybe we just disable `extends`, `subclass` and `dict`? I'll implement this PR as a `#[pyclass]` if you are okay with it.

File: pyo3-macros-backend/src/pyenum.rs
Comment (b05902132) [reply pyo3-macros-backend/src/pyenum.rs:48]: Yes, I already killed that in my local repository.

File: pyo3-macros-backend/src/pyenum.rs
Comment (b05902132) [reply pyo3-macros-backend/src/pyenum.rs:76]: Let me try to figure that out. Meanwhile there's a new problem: I need to implement __eq__ (via __richcmp__) for the enum to be usable, but I'll need a way to do that.

File: src/types/enum.rs
Comment (b05902132) [reply src/types/enum.rs:1]: I didn't see that. Will do.

File: tests/test_enum.rs
Comment (davidhewitt) [reply tests/test_enum.rs:8]: Implementing as `#[pyclass]` sounds great to me!

> Maybe we just disable extends, subclass and dict?

I think `extends` makes sense to disable, however `subclass` and `dict` could reasonably be things that users want to add to customize their enum? I don't think they'll add much complexity to this PR, we can see as we go and decide that later.

File: pyo3-macros-backend/src/pyenum.rs
Comment (davidhewitt) [reply pyo3-macros-backend/src/pyenum.rs:76]: Mmm ultimately you'll need to generate the code to create a `PyType_Slot` struct which points to the implementation of `__richcmp__` you want, which then gets fed into the collection machinery as part of the `for_each_proto_slot` function.

There's some code in `pyo3-macros-backend/src/pymethod.rs` to which handles the magic methods as part of `pymethods` which might be useful if you need to generate a `__richcmp__` method. Or alternatively maybe you're able to hard-code an "enum protocols trait" in `src/class/impl_.rs` which you can then add to enum classes only using the `PyClassImplBuilder`.

Hopefully again those are some notes to get started reading the right code - please do ask more questions as you need!

**Comment by davidhewitt** [comment]
> - How should we treat custom discriminant values? (Add an `__int__` implementation?)

Can you elaborate a little more what you mean by this?

> - How should we implement `__repr__`?
> - How does docstring on each variant work?

I guess that for these two it would make sense to look at how Python does it (maybe Python 3.10?) and aim for a similar end result for users.

File: tests/test_enum.rs
Comment (b05902132) [reply tests/test_enum.rs:8]: I just want to be future-proof. In the future when we add ADT support, each variant will be represented by a subclass (I explained why in [another comment](https://github.com/PyO3/pyo3/issues/417#issuecomment-972623184) of #417), and we want them to be the only subclasses of this enum, otherwise pattern matching will not work. (Python interpreter cannot recognize those subclasses as a variant of the enum.)

I am now okay with enabling `dict`, as I cannot think of a potential reason why we would wan't to disable `dict` in the future.

Comment (davidhewitt) [reply tests/test_enum.rs:8]: > and we want them to be the only subclasses of this enum

Ahh good point, ok yes let's disable `subclass` too.

**Comment by davidhewitt** [comment]
> they will expect some relationship between `Foo.Bar` and `1` at the Python side. I'm not sure what should that be.

Perhaps like Python `IntEnum`, `Foo.Bar.value` should be `1` and also `Foo.Bar == 1`?

File: tests/test_enum.rs
Comment (b05902132) [reply tests/test_enum.rs:8]: Then we just need to figure out how to prevent subclassing from the Python side 😕.

Comment (davidhewitt) [reply tests/test_enum.rs:8]: Worth writing a test but I'm fairly sure that neither Python nor Rust can subclass without `subclass`.

**Comment by b05902132** [comment]
> > they will expect some relationship between `Foo.Bar` and `1` at the Python side. I'm not sure what should that be.
> 
> Perhaps like Python `IntEnum`, `Foo.Bar.value` should be `1` and also `Foo.Bar == 1`?

https://docs.python.org/3/library/enum.html#enum.IntEnum

That's implemented by subclassing `int`... Not sure if we should & how to do that.

I think I'll forbid that for now, and leave that for another PR.

File: tests/test_enum.rs
Comment (b05902132) [reply tests/test_enum.rs:8]: Well, I just want to stop user from writing `#[pyclass(subclass)]`. We still need to secretly enable that in order to write our variant subclasses.

Comment (davidhewitt) [reply tests/test_enum.rs:8]: I see. I'm not sure it can work like that; as far as Python is concerned either it's legal to subclass this or it isn't.

Might be able to control this a bit via a metaclass? Could be overkill.

**Comment by b05902132** [comment]
Hi, I did a major refactoring to implementations of `#[pyclass]`. You might want to review that.

File: tests/test_enum.rs
Comment (b05902132) [reply tests/test_enum.rs]: Python's `Enum` also makes use of metaclass to make sure its enums are immutable. I'll try to implement that after I finish `__eq__`.

File: pyo3-macros-backend/src/pyclass.rs
Comment (b05902132) [inline pyo3-macros-backend/src/pyclass.rs:453]: This is dirty. Have no idea how to change that though.

File: pyo3-macros-backend/src/pyenum.rs
Comment (b05902132) [reply pyo3-macros-backend/src/pyenum.rs]: I have an idea, how about we derive from the enum and provide `__richcmp__` in the subclass? That's less complex, and we don't have to handle the situation where the user want to provide their own `__richcmp__`.
Equivalent code in Python:
```Python
class Enum:
    @classmethod
    @property
    def Var1(cls):
        return EnumVar1()
...

class EnumVar1(Enum):
    def __eq__(self, other):
        # Return true only if the other is a `Enum` and represents Var1
```

File: tests/test_enum.rs
Comment (b05902132) [reply tests/test_enum.rs]: I'm sorry, for some reasons I thought `PyClass::BaseNativeType` must be a class created by the metaclass ... I found I was wrong after reading `create_type_object` and a bunch of CPython code... 😅😅
But I still have no idea whether it's possible create metaclass in PyO3, and if so what are the limitations. Did any one else try this before? Maybe I can do some experiments...

Comment (davidhewitt) [reply tests/test_enum.rs]: Indeed there isn't yet a way to set metaclasses - #906.

Agreed, let's complete this PR without solving the metaclass issue and we can always explore that as a follow-up.

**Comment by davidhewitt** [comment]
👍 will try to re-review in the next couple days when I get a chance!

File: pyo3-macros-backend/src/pyclass.rs
Comment (davidhewitt) [inline pyo3-macros-backend/src/pyclass.rs:486]: There's a `bail_spanned!` macro which can help with boilerplate for this kind of stuff

```suggestion
        _ => bail_spanned!(variant.span() => "Currently only support unit variants.")
```

File: pyo3-macros-backend/src/pyclass.rs
Comment (davidhewitt) [inline pyo3-macros-backend/src/pyclass.rs:391]: Perhaps this name can be more descriptive:

```suggestion
struct PyClassEnumVariant<'a> {
```

File: pyo3-macros-backend/src/pyclass.rs
Comment (davidhewitt) [reply pyo3-macros-backend/src/pyclass.rs:453]: What's your dislike about it? It seems ok to me go through the existing machinery.

**Review by davidhewitt — COMMENTED** [review]
👍  this is looking really solid to me, and nice refactoring work on the `#[pyclass]` implementation.

I'm leaning towards we tidy this PR up just a tiny bit and then merge it, because it's got a lot of changes which might conflict with other work going on.

We can create an issue to record all the pieces of follow-up work to do (such as `__eq__` implementation and guide contents). Hopefully you'd be interested in carrying on with the future work on this? It's an exciting piece of new functionality, thank you for picking this up!

File: pyo3-macros-backend/src/pyclass.rs
Comment (b05902132) [reply pyo3-macros-backend/src/pyclass.rs:453]: I... can't remember 😅 I guess I wanted some constructor function.

**Comment by b05902132** [comment]
> We can create an issue to record all the pieces of follow-up work to do (such as `__eq__` implementation and guide contents).
I'm okay with that, but don't we already have #417 and #834?

> Hopefully you'd be interested in carrying on with the future work on this? It's an exciting piece of new functionality, thank you for picking this up!
I'd love to implement rest of this feature.

File: pyo3-macros-backend/src/pyenum.rs
Comment (davidhewitt) [reply pyo3-macros-backend/src/pyenum.rs]: Perhaps, I think if the user provides their own `__richcmp__` and it overrides our default one that's fine. Let's see how this goes.

**Comment by davidhewitt** [comment]
> I'm okay with that, but don't we already have #417 and #834?

Yes, but I think once this merges we should aim to complete this initial functionality (e.g. `__eq__` and guide entry) before 0.16 releases, so it would be useful to have an issue to remind us of those bits.

File: pyo3-macros-backend/src/pyenum.rs
Comment (davidhewitt) [reply pyo3-macros-backend/src/pyenum.rs]: Another thought could be to enforce singletons for each variant, this removing need for `__richcmp__`

**Review by davidhewitt — APPROVED** [review]
Created #2013 for follow-up work. Post on there if you think anything is missing.

Would you mind rebasing and force-pushing this PR (and potentially squashing it into one or two commits)? The CI seems to have glitched out and so I can't merge it 😅

File: pyo3-macros-backend/src/pyenum.rs
Comment (b05902132) [reply pyo3-macros-backend/src/pyenum.rs:73]: I wanted to use subclass approach because I didn't understand how PyO3 collects methods and thought if we provided `__richcmp__`, the user couldn't create their own. Now I've read the implementation, I know it's not the case and I think I like introducing a new trait. 
I don't know how to enforce singletons.

**Comment by b05902132** [comment]
Rebase done. Sorry there are some hiccups...

**Comment by davidhewitt** [comment]
Brilliant, thanks again!

**Pull request merged**
